src/realm/aggregate_ops.hpp

```diff
@@ -0,0 +1,217 @@
+/*************************************************************************
+ *
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#ifndef REALM_AGGREGATE_OPS_HPP
+#define REALM_AGGREGATE_OPS_HPP
+
+#include <realm/column_type_traits.hpp>
+#include <realm/mixed.hpp>
+
+#include <algorithm>
+
+namespace realm::aggregate_operations {
+
+
+template <class T>
+inline bool valid_for_agg(T)
+{
+    return true;
+}
+
+template <class T>
+inline bool valid_for_agg(util::Optional<T> val)
+{
+    return !!val;
+}
+template <>
+inline bool valid_for_agg(Timestamp val)
+{
+    return !val.is_null();
+}
+inline bool valid_for_agg(StringData val)
+{
+    return !val.is_null();
+}
+inline bool valid_for_agg(BinaryData val)
+{
+    return !val.is_null();
+}
+template <>
+inline bool valid_for_agg(float val)
+{
+    return !null::is_null_float(val) && !std::isnan(val);
+}
+template <>
+inline bool valid_for_agg(double val)
+{
+    return !null::is_null_float(val) && !std::isnan(val);
+}
+template <>
+inline bool valid_for_agg(Decimal128 val)
+{
+    return !val.is_null() && !val.is_nan();
+}
+template <>
+inline bool valid_for_agg(Mixed val)
+{
+    return !val.is_null() && (val.get_type() != type_Decimal || !val.get_decimal().is_nan());
+}
+
+template <typename T, typename Compare>
```

Comment:
These were moved almost verbatim from `src/realm/query_expression.hpp` so that they could be shared between queries and object accessor aggregates. Previously we had two implementations which were slightly different in handling null/NaN.